src/pip/_internal/exceptions.py

```diff
@@ -28,6 +29,36 @@ class UninstallationError(PipError):
     """General exception during uninstallation"""
 
 
+class NoneMetadataError(PipError):
+    """
+    Raised when accessing "METADATA" or "PKG-INFO" metadata for a
+    pip._vendor.pkg_resources.Distribution object and
+    `dist.has_metadata('METADATA')` returns True but
+    `dist.get_metadata('METADATA')` returns None (and similarly for
+    "PKG-INFO").
+    """
+
+    def __init__(self, dist, metadata_name):
+        # type: (Distribution, str) -> None
+        """
+        :param dist: A Distribution object.
+        :param metadata_name: The name of the metadata being accessed
+            (can be "METADATA" or "PKG-INFO").
+        """
+        self.dist = dist
+        self.metadata_name = metadata_name
+
+    def __str__(self):
+        # type: () -> str
+        # Use `dist` in the error message because its stringification
+        # includes more information, like the version and location.
+        return (
+            'None {} metadata found for distribution: {}'.format(
```

Comment:
"No metadata found for distribution: {dist}" seems enough to me.

The distribution usually contains egg-info or dist-info extension on the string representation, right? If not, I'd prefer to keep the file name after the distribution, since that's more of a debugging detail IMO.

Reply:
> "No metadata found for distribution: {dist}" seems enough to me.

There's a distinction here the message is trying to preserve. `dist.has_metadata()` can return `False`, which is one form of "no metadata," but the case of this exception is `dist.has_metadata()` True but the return value being `None`. The former case doesn't currently raise an exception and is being logged as, "No metadata found in ...". The latter is a very specific kind of error (and maybe shouldn't ever be happening).

> The distribution usually contains egg-info or dist-info extension on the string representation, right?

I checked, and no, it doesn't include that. Also, I see that `dist.location` isn't even necessarily included, so I should probably add that.

Reply:
Thanks for taking a look. Your rationale sounds good to me.

Reply:
Okay, great. And thanks a lot for reviewing this quickly.

Reply:
No worries. Thanks for filing such polished PRs.

Reply:
By the way, the exception approach you liked was I think inspired subconsciously by this issue you had filed earlier: https://github.com/pypa/pip/issues/6541 👍